d2layouts/d2sequence/layout.go

```diff
@@ -4,90 +4,140 @@ import (
 	"context"
 	"fmt"
 	"math"
+	"sort"
 
 	"oss.terrastruct.com/d2/d2graph"
 	"oss.terrastruct.com/d2/lib/geo"
 	"oss.terrastruct.com/d2/lib/go2"
 	"oss.terrastruct.com/d2/lib/label"
+	"oss.terrastruct.com/d2/lib/shape"
 )
 
 func Layout(ctx context.Context, g *d2graph.Graph) (err error) {
-	edgeYStep := MIN_EDGE_DISTANCE
-	actorXStep := MIN_ACTOR_DISTANCE
-	maxActorHeight := 0.
+	sd := &sequenceDiagram{
+		graph:          g,
+		objectRank:     make(map[*d2graph.Object]int),
+		objectDepth:    make(map[*d2graph.Object]int),
+		minEdgeRank:    make(map[*d2graph.Object]int),
+		maxEdgeRank:    make(map[*d2graph.Object]int),
+		edgeYStep:      MIN_EDGE_DISTANCE,
+		actorXStep:     MIN_ACTOR_DISTANCE,
+		maxActorHeight: 0.,
+	}
+
+	sd.init()
+	sd.placeActors()
+	sd.placeSpanBoxes()
+	sd.routeEdges()
+	sd.addLifelineEdges()
+
+	return nil
+}
+
+type sequenceDiagram struct {
+	graph *d2graph.Graph
+
+	edges  []*d2graph.Edge
+	actors []*d2graph.Object
+	spans  []*d2graph.Object
+
+	// can be either actors or span boxes
+	// rank: left to right position of actors/span boxes
```

Comment:
u mean index?

Reply:
To some extent.
Like, `spans` have the same rank as their parents.
Index gets some ambiguity with the array where they are stored.